staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -132,24 +131,39 @@ func NonSlidingUntilWithContext(ctx context.Context, f func(context.Context), pe
 // Close stopCh to stop. f may not be invoked if stop channel is already
 // closed. Pass NeverStop to if you don't want it stop.
 func JitterUntil(f func(), period time.Duration, jitterFactor float64, sliding bool, stopCh <-chan struct{}) {
-	BackoffUntil(f, NewJitteredBackoffManager(period, jitterFactor, &clock.RealClock{}), sliding, stopCh)
+	b := Backoff{Duration: period, Jitter: jitterFactor}
+	BackoffUntil(f, RealTimer, b.Step, sliding, stopCh)
 }
 
+var (
+	// RealTimer can be passed to methods that need a TimerFunc.
+	RealTimer = clock.RealClock{}.NewTimer
+)
+
+// TimerFunc allows the caller to inject a timer for testing. Most callers should use RealTimer.
+type TimerFunc func(time.Duration) clock.Timer
+
 // BackoffUntil loops until stop channel is closed, run f every duration given by BackoffManager.
 //
 // If sliding is true, the period is computed after f runs. If it is false then
 // period includes the runtime for f.
-func BackoffUntil(f func(), backoff BackoffManager, sliding bool, stopCh <-chan struct{}) {
+func BackoffUntil(f func(), timerFn TimerFunc, delayFn DelayFunc, sliding bool, stopCh <-chan struct{}) {
 	var t clock.Timer
+	defer func() {
+		if t != nil {
+			t.Stop()
+		}
+	}()
```

Comment:
wow, this was a subtle bug, if the previous code returned on the first select it will never stop the timer

Reply:
It *should* lead to a goroutine leak, but i'm not 100% positive.  We didn't seem to see signs of it but we also don't use a ton of wait loops that exit except at cleanup.